internal/entitlement/balance_connector.go

```diff
@@ -0,0 +1,304 @@
+	"github.com/openmeterio/openmeter/internal/streaming"
+	"github.com/openmeterio/openmeter/pkg/convert"
+	"github.com/openmeterio/openmeter/pkg/models"
+	"github.com/openmeterio/openmeter/pkg/slicesx"
+)
+
+type EntitlementBalance struct {
+	EntitlementID string    `json:"entitlementId"`
+	Balance       float64   `json:"balance"`
+	UsageInPeriod float64   `json:"usageInPeriod"`
+	Overage       float64   `json:"overage"`
+	StartOfPeriod time.Time `json:"startOfPeriod"`
+}
+
+type EntitlementBalanceHistoryWindow struct {
+	From           time.Time
+	To             time.Time
+	UsageInPeriod  float64
+	BalanceAtStart float64
+	OverageAtStart float64
+}
+
+type WindowSize string
+
+const (
+	// We don't support minute precision as that results in an extremely heavy calculation
+	// WindowSizeMinute WindowSize = "MINUTE"
+	WindowSizeHour WindowSize = "HOUR"
+	WindowSizeDay  WindowSize = "DAY"
+)
+
+type BalanceHistoryParams struct {
+	From           time.Time
+	To             time.Time
+	WindowSize     WindowSize
+	WindowTimeZone time.Location
+}
+
+type EntitlementBalanceConnector interface {
+	GetEntitlementBalance(ctx context.Context, entitlementID models.NamespacedID, at time.Time) (*EntitlementBalance, error)
+	GetEntitlementBalanceHistory(ctx context.Context, entitlementID models.NamespacedID, params BalanceHistoryParams) ([]EntitlementBalanceHistoryWindow, credit.GrantBurnDownHistory, error)
+	ResetEntitlementUsage(ctx context.Context, entitlementID models.NamespacedID, resetAt time.Time) (balanceAfterReset *EntitlementBalance, err error)
+
+	// GetEntitlementGrantBalanceHistory(ctx context.Context, entitlementGrantID EntitlementGrantID, params BalanceHistoryParams) ([]EntitlementBalanceHistoryWindow, error)
+	CreateGrant(ctx context.Context, entitlement models.NamespacedID, inputGrant CreateEntitlementGrantInputs) (EntitlementGrant, error)
+	ListEntitlementGrants(ctx context.Context, entitlementID models.NamespacedID) ([]EntitlementGrant, error)
+}
+
+type entitlementBalanceConnector struct {
+	sc streaming.Connector
+	oc credit.OwnerConnector
+	bc credit.BalanceConnector
+	gc credit.GrantConnector
+}
+
+func NewEntitlementBalanceConnector(
+	sc streaming.Connector,
+	oc credit.OwnerConnector,
+	bc credit.BalanceConnector,
+	gc credit.GrantConnector,
+) EntitlementBalanceConnector {
+	return &entitlementBalanceConnector{
+		sc: sc,
+		oc: oc,
+		bc: bc,
+		gc: gc,
+	}
+}
+
+func (e *entitlementBalanceConnector) GetEntitlementBalance(ctx context.Context, entitlementID models.NamespacedID, at time.Time) (*EntitlementBalance, error) {
+	nsOwner := credit.NamespacedGrantOwner{
+		Namespace: entitlementID.Namespace,
+		ID:        credit.GrantOwner(entitlementID.ID),
+	}
+	res, err := e.bc.GetBalanceOfOwner(ctx, nsOwner, at)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get balance of entitlement %s: %w", entitlementID.ID, err)
+	}
+
+	meterSlug, params, err := e.oc.GetOwnerQueryParams(ctx, nsOwner)
```

Comment:
If I understand correctly, during `e.bc.GetBalanceOfOwner`, we are relying on the meter data to calculate grant burndown and overage. Thus, we have the usage data there, too.

Question: given that we already have that data queried there, can't we just return the usage as part of the balance information?

(This would allow this connector to be a really thin one, as I think the history has the same issue).

Reply:
we can't: usage is measured for the whole period while the Engine only calculates since the last snapshot